Emails: Logic to show either inbox selector or domains picker or domains CTA buy

AllTerrainDeveloper opened the pull request:
#### Changes proposed in this Pull Request

This pull request adds the logic to show in upgrade->emails mailboxes if we have mailboxes to show. It will show either add a domain CTA or select domain if we have available ones.

#### Testing instructions

Have the following sites ready:
Site with no domains, site with one domain and no email subscription, sites with more than one domain with email subscriptions, and a site with mailboxes configured

**Test 1, Site with no domains:**
1. Go to the url: /inbox/{your-site-slug-with-no-domains}
2. Check that the inbox management screen is presenting you with a CTA to buy a domain

![image](https://user-images.githubusercontent.com/5689927/134048948-3db4d443-9997-4390-9c27-b079149058c1.png)

**Test 2, site with one domain and no email subscription**
1. Go to the url: /inbox/{your-site-slug-with-domain-and-no-email}
2. Check that the inbox management screen is presenting you with email comparison page

![image](https://user-images.githubusercontent.com/5689927/134049602-ee938ff5-db4b-4610-84d8-09ea54e79d65.png)

**Test 3, have a site with several domains without subscriptions**
1. Go to the url: /inbox/{your-site-with-several-domains-and-no-emails}
2. Check that the inbox management screen is presenting you with a domain selector, that takes you to email comparison page.

![image](https://user-images.githubusercontent.com/5689927/134499928-c4dbc539-a004-4d28-b37f-9b4d1a041750.png)

**Test 4, have a site with minimum a domain with subscriptions**
1. Go to the url: /inbox/{your-site-with-at-least-a-domain-and-emails}
2. Check that the inbox management screen is presenting you with a placeholder screen, this screen will be developed shortly.

#### Regression Test

The component `EmailListInactive` has been "upgraded" to change it's wording when it is being used in the mailbox component. Also, the component `EmailsHome` has been updated to do not show active domains (domains with subscriptions) when it is presented from inbox management screen. We need to check that the component `EmailsHome`  keeps working as it was doing it before:

BEFORE | AFTER
------------ | -------------
![image](https://user-images.githubusercontent.com/5689927/134046870-994aa932-ddb9-4da2-8d10-a4fb46e2cb17.png) | ![image](https://user-images.githubusercontent.com/5689927/134048264-5c98b25f-7fbf-4625-9458-eb293e3f5142.png)

Related to {1200182182542585-as-1200879615791996}

AllTerrainDeveloper committed:
Emails: Logic to show either inbox selector or domains picker or domains CTA buy.

AllTerrainDeveloper committed:
Emails: Improved logic and removed active emails (with subscriptions) from inbox management screen

AllTerrainDeveloper committed:
Emails: Better logic

stephanethomas commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -117,14 +118,18 @@ class EmailManagementHome extends React.Component {
-					currentRoute={ currentRoute }
-					selectedSiteId={ selectedSiteId }
-				/>
+				{ emailListActive }
```
Not mandatory but we could keep that component inlined like this:

```suggestion
				{ isUnderEmailInbox( currentRoute ) && <EmailListActive ... }
```

AllTerrainDeveloper replied:
Yes, I like it more what you said, we can see at a glance what is everything.

stephanethomas commented on client/my-sites/email/email-management/home/email-list-inactive.jsx:
```diff
@@ -1,18 +1,54 @@
+		return (
+			<>
+				<PromoCard
+					isPrimary={ true }
```
We don't need to specify `true` here, it can be omitted (this is the default value for props not specified):

```suggestion
					isPrimary
```

AllTerrainDeveloper replied:
Actually, it was needed :D the default value was undefined.

stephanethomas replied:
`PromoCard` is a TypeScript component, maybe that's why. I had a look to the documentation, and interestingly it is now [discouraged to omit that value](https://reactjs.org/docs/jsx-in-depth.html#props-default-to-true). Good to know!

stephanethomas commented on client/my-sites/email/email-management/home/email-list-inactive.jsx:
```diff
@@ -1,18 +1,54 @@
+				>
+					<p>
+						{ translate(
+							'Pick a domain from your available domains below to add email service to it.'
```
I'm not sure if that's normal but the mockup had a slightly different copy:

```suggestion
							'Pick a domain from your available domains below to add an email solution.'
```

stephanethomas commented on client/my-sites/email/inbox-management/index.js:
```diff
@@ -0,0 +1,36 @@
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+const InboxManagement = ( { domains } ) => {
+	const nonWPCOMDomains = domains.filter( ( domain ) => ! domain.isWPCOMDomain );
+	const domainsWithSubscriptions = nonWPCOMDomains.filter( ( domain ) =>
+		hasEmailSubscription( domain )
+	);
```
It looks like we could simplify this block of code like this:

```suggestion
	const domainsWithSubscriptions = domains.filter( ( domain ) => ! domain.isWPCOMDomain && hasEmailSubscription( domain ) );
```

AllTerrainDeveloper replied:
Right!

stephanethomas commented on client/my-sites/email/paths.js:
```diff
@@ -171,11 +172,15 @@ export function emailManagementEdit(
 	return path?.startsWith( emailManagementAllSitesPrefix + '/' );
 }
+
+export function isUnderEmailInbox( path ) {
```
Shouldn't we use a name that better matches the path?

```suggestion
export function isUnderEmailManagementInbox( path ) {
```

AllTerrainDeveloper replied:
Yes, thank you :)

AllTerrainDeveloper committed:
Emails: PR suggestions - Rename functions, inline functions, refactor code

AllTerrainDeveloper committed:
Emails: Fix LINT problems

daledupreez commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -43,6 +46,7 @@ class EmailManagementHome extends React.Component {
 		selectedDomainName: PropTypes.string,
 		selectedSiteId: PropTypes.number.isRequired,
 		selectedSiteSlug: PropTypes.string.isRequired,
+		context: PropTypes.string,
```
Thanks for catching this!

daledupreez commented on client/my-sites/email/email-management/home/email-list-inactive.jsx:
```diff
@@ -1,18 +1,50 @@
+	}
+
 	render() {
 		const { selectedSiteSlug, currentRoute, domains, translate } = this.props;
 
 		if ( domains.length < 1 ) {
 			return null;
 		}
 
+		const sectionHeaderLabel = isUnderEmailManagementInbox( currentRoute )
+			? translate( 'Domains' )
+			: translate( 'Other domains' );
+
+		const mainHeader = isUnderEmailManagementInbox( currentRoute ) ? this.getMainHeader() : null;
+
```
I would prefer it if this logic was handled by props, where we can push this logic onto the caller - it doesn't feel right to me for this component to need to know what context it is running in.
```suggestion
		const { currentRoute, domains, header, sectionLabel, selectedSiteSlug, translate } = this.props;

		if ( domains.length < 1 ) {
			return null;
		}

		const sectionHeaderLabel = sectionLabel ? sectionLabel : translate( 'Other domains' );

```

AllTerrainDeveloper replied:
I have reworked this but added modifications in the caller (email-home) to pass those properties to the child (email-list-inactive).

Flow is that inbox-management calls email-home with parameters and email-home calls email-list-inactive with those parameters.

daledupreez reviewed (changes requested):
Overall, this is looking good, but I tripped over a CSS issue with the new promo card/header for the case where we have multiple domains without subscriptions.

<img width="1090" alt="Screenshot 2021-09-21 at 21 16 57" src="https://user-images.githubusercontent.com/3376401/134234617-9ee6a17b-e8da-4042-bf09-2b1c064d3312.png">

It looks like we may need some additional CSS to override the default CSS for the image within the card.

daledupreez commented on client/my-sites/email/email-management/home/email-list-inactive.jsx:
```diff
@@ -1,18 +1,50 @@
+				<PromoCard title={ translate( 'Pick a domain to get started' ) } image={ image }>
+					<p>
+						{ translate(
+							'Pick a domain from your available domains below to add an email solution.'
```
How about this simpler wording as well? cc @saygunnyc 
```suggestion
							'Pick one of your domains below to add an email solution.'
```

AllTerrainDeveloper replied:
In the meanwhile, I have added your suggestion, if Saygun thinks that it's better the other wording I'll change it :)

AllTerrainDeveloper committed:
Emails: Injected props instead of having a component that knows what to show depending on the current route.

AllTerrainDeveloper committed:
Emails: Section header label in email-list-inactive fixed

AllTerrainDeveloper committed:
Emails: Fix default label

AllTerrainDeveloper commented:
@daledupreez I have reworked the way that the components are working. Now they are just receiving by props how they should behave. And the CSS thing you noticed, good catch! Wasn't actually a CSS fix, I removed a property from PromoCard that was necessary => `is-primary = true`

AllTerrainDeveloper committed:
Emails: Lint

olaseni commented on client/my-sites/email/email-management/home/email-list-inactive.jsx:
```diff
@@ -7,8 +7,14 @@ import { emailManagementPurchaseNewEmailAccount } from 'calypso/my-sites/email/p
+		const {
+			currentRoute,
+			domains,
+			header,
```
It might be worth adding proptypes for these.

olaseni replied:
Also, would this be better named as `headerComponent` seeing as it's actually a component prop ?

olaseni commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -43,18 +46,25 @@ class EmailManagementHome extends React.Component {
 		selectedSiteSlug: PropTypes.string.isRequired,
+		context: PropTypes.string,
+		emailListInactiveHeader: PropTypes.element,
+		onlyDomainsWithoutSubscription: PropTypes.bool,
```
Would it be simpler to name this `showActiveDomainList` ?

olaseni reviewed (changes requested):
I left a couple of suggestions on readability. You also need to rebase to pull in relevant changes.

saygunnyc commented:
@AllTerrainDeveloper Let's make sure we're using the same CSS properties from the existing email comparison page including the size & location of the illustration. The screenshots shared above seem different.

Select A Domain Page
<img width="733" alt="Screen Shot 2021-09-22 at 15 20 48" src="https://user-images.githubusercontent.com/78876323/134342582-b6abbbce-c847-428a-aa84-572053331406.png">
 
Email Comparison Page
<img width="738" alt="Screen Shot 2021-09-22 at 15 20 45" src="https://user-images.githubusercontent.com/78876323/134342578-e5c6624c-cf9d-453a-b91b-31e75ef951a0.png">

AllTerrainDeveloper committed:
Emails: Fix styling and new rename variables

AllTerrainDeveloper commented:
> @AllTerrainDeveloper Let's make sure we're using the same CSS properties from the existing email comparison page including the size & location of the illustration. The screenshots shared above seem different.
> 
> Select A Domain Page
> <img alt="Screen Shot 2021-09-22 at 15 20 48" width="733" src="https://user-images.githubusercontent.com/78876323/134342582-b6abbbce-c847-428a-aa84-572053331406.png">
> 
> Email Comparison Page
> <img alt="Screen Shot 2021-09-22 at 15 20 45" width="738" src="https://user-images.githubusercontent.com/78876323/134342578-e5c6624c-cf9d-453a-b91b-31e75ef951a0.png">

Thanks Saygun! Good catch, I've updated the Pull Request description to match it after I have fixed this issue :)

AllTerrainDeveloper committed:
Emails: Lint

olaseni commented on client/my-sites/email/email-management/email-home.jsx:
```diff
@@ -119,16 +126,20 @@ class EmailManagementHome extends React.Component {
-					currentRoute={ currentRoute }
-					selectedSiteId={ selectedSiteId }
-				/>
+				{ ! showActiveDomainList && (
```
This should be:
```suggestion
				{ showActiveDomainList && (
```

You could also set a default prop value of `true`

AllTerrainDeveloper committed:
Emails: Change meaning of variable showActiveDomainList

olaseni reviewed (approved):
This is looking real good. Let's do it.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6648463

Thank you @AllTerrainDeveloper for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6648463).
